src/deps-index.ts

```diff
@@ -0,0 +1,99 @@
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Analyzer} from 'polymer-analyzer';
+
+export interface DepsIndex {
+  // An index of dependency -> fragments that depend on it
+  depsToFragments: Map<string, Set<string>>;
+  // An index of fragments -> html dependencies
+  fragmentToDeps: Map<string, Set<string>>;
+}
+
+export interface BundleManifest extends Map<string, Set<string>> {}
+
+export interface BundleOrchestrator {
+  (endpoints: string[], depsIndex: DepsIndex): BundleManifest
+}
+
+type DependencyMapEntry = {
+  url: string,
+  eager: Set<string>,
+  lazy: Set<string>
+};
+
+async function _getDependencies(href: string, analyzer: Analyzer):
+    Promise<DependencyMapEntry> {
+      const document = await analyzer.analyzeRoot(href);
+      const imports = document.getByKind('import');
+      const eagerImports = new Set<string>();
+      const lazyImports = new Set<string>();
+      for (let htmlImport of imports) {
+        if (!htmlImport.url) {
+          continue;
+        }
+        switch (htmlImport.type) {
+          case 'html-import':
+            eagerImports.add(htmlImport.url);
+            break;
+          case 'lazy-html-import':
+            lazyImports.add(htmlImport.url);
+            break;
+        }
+      }
+      return {url: href, eager: eagerImports, lazy: lazyImports};
+    }
+
+export async function buildDepsIndex(entrypoints: string[], analyzer: Analyzer):
+    Promise<DepsIndex> {
+  const entrypointToDependencies: Map<string, Set<string>> = new Map();
+  const dependenciesToEntrypoints: Map<string, Set<string>> = new Map();
+  let depsIndex = {};
+  const queue = entrypoints.slice();
+  const visitedEntrypoints = new Set<string>();
+  while (queue.length > 0) {
+    const entrypoint = queue.shift()!;
+    if (visitedEntrypoints.has(entrypoint)) {
+      continue;
+    }
+    const dependencyEntry = await _getDependencies(entrypoint, analyzer);
+    const dependencies = new Set(dependencyEntry.eager);
+    dependencies.add(entrypoint);
+    entrypointToDependencies.set(entrypoint, dependencies);
+    for (const lazyDependency of dependencyEntry.lazy.values()) {
+      if (!visitedEntrypoints.has(lazyDependency)) {
+        queue.push(lazyDependency);
+      }
+    }
+  }
+  entrypointToDependencies.forEach((dependencies, entrypoint, map) => {
+    for (const dependency of dependencies) {
+      if (!dependenciesToEntrypoints.has(dependency)) {
+        dependenciesToEntrypoints.set(dependency, new Set());
+      }
+      const entrypointSet = dependenciesToEntrypoints.get(dependency)!;
+      entrypointSet.add(entrypoint);
+    }
+  });
+  return {
+    depsToFragments: dependenciesToEntrypoints,
+    fragmentToDeps: entrypointToDependencies
+  };
+}
+
+
+export function defaultHeuristic(entrypoints: string, index: DepsIndex):
```

Comment:
Let's name the built-in heuristics. I think we should have at least one-bundle and maximum sharding (the identity function on the bundle manifest).